### PR TITLE
AIRFLOW-5581: Join KubernetesJobWatcher in terminate call and unblock queues from blocking forever

### PR DESCRIPTION
**Pull request by kpathak13**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR:
  - https://issues.apache.org/jira/browse/AIRFLOW-5581

### Description

- [x] It fixes a bug in `kubernetes_executor` module

> If run_duration or num_runs is enabled, KubernetesJobWatcher dies prematurely causing below error:
>
```text
[2019-10-01 19:40:35,323] {kubernetes_executor.py:327} ERROR - Unknown error in KubernetesJobWatcher. Failing
Traceback (most recent call last):
File "/home/vccorp/.local/lib/python3.6/site-packages/airflow/contrib/executors/kubernetes_executor.py", line 325, in run
self.worker_uuid, self.kube_config)
File "/home/vccorp/.local/lib/python3.6/site-packages/airflow/contrib/executors/kubernetes_executor.py", line 359, in _run
task.metadata.resource_version
File "/home/vccorp/.local/lib/python3.6/site-packages/airflow/contrib/executors/kubernetes_executor.py", line 391, in process_status
self.watcher_queue.put((pod_id, None, labels, resource_version))
File "", line 2, in put
File "/usr/lib64/python3.6/multiprocessing/managers.py", line 756, in _callmethod
conn.send((self._id, methodname, args, kwds))
File "/usr/lib64/python3.6/multiprocessing/connection.py", line 206, in send
self._send_bytes(_ForkingPickler.dumps(obj))
File "/usr/lib64/python3.6/multiprocessing/connection.py", line 404, in _send_bytes
self._send(header + buf)
File "/usr/lib64/python3.6/multiprocessing/connection.py", line 368, in _send
n = write(self._handle, buf)
BrokenPipeError: [Errno 32] Broken pipe
Process KubernetesJobWatcher-3:
Traceback (most recent call last):
File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
self.run()
File "/home/vccorp/.local/lib/python3.6/site-packages/airflow/contrib/executors/kubernetes_executor.py", line 325, in run
self.worker_uuid, self.kube_config)
File "/home/vccorp/.local/lib/python3.6/site-packages/airflow/contrib/executors/kubernetes_executor.py", line 359, in _run
task.metadata.resource_version
File "/home/vccorp/.local/lib/python3.6/site-packages/airflow/contrib/executors/kubernetes_executor.py", line 391, in process_status
self.watcher_queue.put((pod_id, None, labels, resource_version))
File "", line 2, in put
File "/usr/lib64/python3.6/multiprocessing/managers.py", line 756, in _callmethod
conn.send((self._id, methodname, args, kwds))
File "/usr/lib64/python3.6/multiprocessing/connection.py", line 206, in send
self._send_bytes(_ForkingPickler.dumps(obj))
File "/usr/lib64/python3.6/multiprocessing/connection.py", line 404, in _send_bytes
self._send(header + buf)
File "/usr/lib64/python3.6/multiprocessing/connection.py", line 368, in _send
n = write(self._handle, buf)
BrokenPipeError: [Errno 32] Broken pipe
```

### Tests

- [x] This does not have test case since it's a single line update.

### CODE REVIEW

File: airflow/executors/kubernetes_executor.py [hunk 1]
```diff
@@ -576,6 +576,7 @@ def _labels_to_key(self, labels):
     def terminate(self):
         """Termninates the watcher."""
         self.watcher_queue.join()
+        self.kube_watcher.join()
```
Comment (ashb) [inline airflow/executors/kubernetes_executor.py:590]: I don't think this is right. This doesn't signal in any way to the kube_watcher that it should shutdown - it simply blocks execution until it stops.

Comment (kpathak13) [reply airflow/executors/kubernetes_executor.py:590]: Correct me if I am wrong @ashb, I believe this call [self.processor_agent.terminate()](https://github.com/apache/airflow/blob/master/airflow/jobs/scheduler_job.py#L1434) signals the `manager` to terminate.

Comment (ashb) [reply airflow/executors/kubernetes_executor.py:590]: No not at all. This has nothing to do with the scheduler_job

Comment (kpathak13) [reply airflow/executors/kubernetes_executor.py:590]: Thanks @ashb. I have added `terminate` call.

File: airflow/executors/kubernetes_executor.py [hunk 2]
```diff
@@ -768,9 +787,45 @@ def _change_state(self, key, state, pod_id: str) -> None:
...
+                finally:
+                    self.result_queue.task_done()
+            except Empty:
+                break
+
     def end(self):
         """Called when the executor shuts down"""
         self.log.info('Shutting down Kubernetes executor')
+        self.log.debug('Flushing task_queue...')
+        self._flush_task_queue()
```
Comment (dimberman) [inline airflow/executors/kubernetes_executor.py:825]: Is there any chance of a race condition where it continues putting in new tasks after the queue has been flushed?

**Review by dimberman — CHANGES_REQUESTED** [review]

File: airflow/executors/kubernetes_executor.py (hunk 2 above)
Comment (kpathak13) [reply airflow/executors/kubernetes_executor.py:825]: That should not be the case since `executor.end()` is called after exiting the executor loop.

**Comment by dimberman** [comment]
@kpathak13 LGTM. Can you rebase? I'll merge when ready :)

**Review by dimberman — APPROVED** [review]

File: airflow/executors/kubernetes_executor.py [hunk 3]
```diff
@@ -572,9 +572,28 @@ def _labels_to_key(self, labels):
...
+                self.watcher_queue.task_done()
+            except Empty:
+                break
+
     def terminate(self):
         """Termninates the watcher."""
+        self.log.debug("Terminating kube_watcher...")
+        self.kube_watcher.terminate()
+        self.kube_watcher.join()
+        self.log.debug("kube_watcher=%s alive?=%s", self.kube_watcher, self.kube_watcher.is_alive())
```
Comment (ashb) [inline airflow/executors/kubernetes_executor.py]: This isn't useful - after a call to `.join()` the kube_watcher should never be alive, right?

Comment (kpathak13) [reply airflow/executors/kubernetes_executor.py]: Removed `is alive?` and rebased.

**Comment by ashb** [comment]
The code looks okay, but the important question: how have you tested this?

**Comment by kpathak13** [comment]
@ashb , I tested this on my k8 cluster:  

    1. Scheduled multiple high frequency jobs 
    2. Set Scheduler run duration to 2 minutes
    3. Killed several pods in flight at different stages (multiple times)
    4. Deleted Scheduler pod

Never observed race condition.

**Review by ashb — APPROVED** [review]

**Pull request merged**